### PR TITLE
added support for new PD

### PR DESCRIPTION
**Pull request by shakyShane**
**Reviewer:** @jonathanKingston 

https://app.asana.com/0/0/1203280236620710/f

## Description:
Added support for new Privacy Dashboard

## Automated tests:
- [x] Unit tests
- [x] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [x] Get advice or leverage existing code
- [x] Agree on technical approach with reviewer (if the changes are nuanced)
- [x] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

**Comment by shakyShane** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #1271** <a href="https://app.graphite.dev/github/pr/duckduckgo/duckduckgo-privacy-extension/1271" target="_blank">Graphite</a>  👈
    * **PR #1272** <a href="https://app.graphite.dev/github/pr/duckduckgo/duckduckgo-privacy-extension/1272" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/duckduckgo-privacy-extension/1271?utm_source=stack-comment).

File: shared/js/background/message-handlers.js [hunk 1]
```diff
@@ -65,12 +66,38 @@ export function getBrowser () {
...
+ */
+async function getTabOriginal (tabId) {
+    // Await for storage to be ready; this happens on service worker closing mostly.
+    await settings.ready()
+    await tdsStorage.ready('config')
+
+    return tabManager.getOrRestoreTab(tabId)
 }
 
 export async function getTab (tabId) {
```
Comment (shakyShane) [inline shared/js/background/message-handlers.js:103]: double-check if this is used elsewhere?

File: shared/js/background/message-handlers.js [hunk 2]
```diff
@@ -65,12 +66,38 @@ export function getBrowser () {
...
+    }
+
+    return breakageReportForTab(tab, category, description)
+}
+
+/**
+ * @param tabId
+ * @returns {Promise<import("./classes/tab.es6")>}
+ */
+async function getTabOriginal (tabId) {
```
Comment (shakyShane) [inline shared/js/background/message-handlers.js:95]: move this util out of message-handlers

**Comment by shakyShane** [comment]
@jonathanKingston 1 quick thing:

1 - there a bug with searching at the moment (the field gets wiped when new data arrives) - it'll be fixed soon and doesn't affect any changes here

File: integration-test/background/privacy-dashboard.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    it('Should load the dashboard with correct link text', async () => {
+        // Load the test page.
+        const page = await browser.newPage()
+        await pageWait.forGoto(page, testSite)
+
+        await page.click('#start')
+
+        const panelUrl = await bgPage.evaluate(async () => {
+            const currentTab = await globalThis.dbg.utils.getCurrentTab()
+            return chrome.runtime.getURL(`dashboard/html/browser.html?tabId=${currentTab.id}`)
```
Comment (jonathanKingston) [inline integration-test/background/privacy-dashboard.js:30]: Nice thanks for adding this!

File: shared/js/background/classes/tab.es6.js [hunk 4]
```diff
@@ -214,47 +201,30 @@ class Tab {
...
+     * @param {string} url
+     * @returns {Tracker}
+     */
+    addToTrackers (t, baseDomain, url) {
         const trackers = this.trackers
         const tracker = this.trackers[t.tracker.owner.name]
 
         if (tracker) {
-            tracker.addTrackerUrl(t)
+            tracker.addTrackerUrl(t, this.url || '', baseDomain, url)
```
Comment (jonathanKingston) [inline shared/js/background/classes/tab.es6.js:223]: I think we could probably clean this up to not need the `t.tracker.owner.name` keying here but let's leave that to a follow up.

File: shared/js/background/message-handlers.js [hunk 5]
```diff
@@ -65,32 +65,60 @@ export function getBrowser () {
...
+/**
+ * This message is here to ensure the privacy dashboard can render
+ * from a single call to the extension.
+ *
+ * Currently, it will collect data for the current tab and email protection
+ * user data.
+ */
+export async function getPrivacyDashboardData (options) {
+    let { tabId } = options
+    if (Number(tabId) === 0) {
```
Comment (jonathanKingston) [inline shared/js/background/message-handlers.js:110]: It seems a little risky to use 0th tab id's can we use null instead?

File: shared/js/ui/base/ui-wrapper.es6.js [hunk 6]
```diff
@@ -18,37 +16,7 @@ export const backgroundMessage = (thisModel) => {
...
-    }
-    if (tabId) {
-        const backgroundTabObj = await sendMessage('getTab', tabId)
-        if (backgroundTabObj) {
-            return backgroundTabObj
-        }
-    }
-}
-
-export const search = (url) => {
```
Comment (jonathanKingston) [inline shared/js/ui/base/ui-wrapper.es6.js:47]: I'm guessing this should be an message-handler now?

File: shared/js/background/message-handlers.js [hunk 7]
```diff
@@ -65,32 +65,60 @@ export function getBrowser () {
...
+ * @returns {Promise<import("./classes/tab.es6")>}
+ */
+async function getTab (tabId) {
     // Await for storage to be ready; this happens on service worker closing mostly.
     await settings.ready()
     await tdsStorage.ready('config')
-
-    const tab = await tabManager.getOrRestoreTab(tabId)
-    return new LegacyTabTransfer(tab)
+    return tabManager.getOrRestoreTab(tabId)
```
Comment (jonathanKingston) [inline shared/js/background/message-handlers.js:98]: `getOrRestoreTab` perhaps could have these two things to wait on instead; I think the `getTab` message handler isn't used now right?

File: shared/js/background/classes/tracker.js [hunk 8]
```diff
@@ -43,22 +29,44 @@ export class Tracker {
...
+        /** @type {DetectedRequestWithAction} */
+        const detectedRequest = {
+            action: t.action,
+            url,
+            eTLDplus1: baseDomain,
+            pageUrl: tabUrl,
+            entityName: this.displayName,
+            prevalence: this.prevalence,
+            ownerName: this.parentCompany?.name,
+            category: t.tracker?.categories?.find(category => constants.displayCategories.includes(category)),
```
Comment (jonathanKingston) [inline shared/js/background/classes/tracker.js:65]: This line has enough complexity to be it's own const.
```js
const category = t.tracker?.categories?.find(category => constants.displayCategories.includes(category))
```

File: shared/js/background/classes/tracker.js [hunk 9]
```diff
@@ -1,36 +1,22 @@
...
+import { convertState } from './privacy-dashboard-data'
+
 const Companies = require('../companies.es6')
 const tdsStorage = require('../storage/tds.es6')
 
 /**
  * @typedef {import('../trackers.es6').ActionName} ActionName
  * @typedef {import('../trackers.es6').TrackerData} TrackerData
+ * @typedef {import('@duckduckgo/privacy-dashboard/schema/__generated__/schema.types').DetectedRequest} DetectedRequest
+ * @typedef {DetectedRequest & { action: ActionName }} DetectedRequestWithAction
```
Comment (jonathanKingston) [inline shared/js/background/classes/tracker.js:11]: 👍🏻

File: shared/js/background/classes/privacy-dashboard-data.js [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+function convertToRequests (tab, protectionsEnabled) {
+    /** @type {DetectedRequest[]} */
+    const detectedRequests = []
+    for (const tracker of Object.values(tab.trackers || {})) {
+        for (const detectedRequest of Object.values(tracker.urls || {})) {
+            // When protections are off, change the 'state' of each tracking request
+            if (!protectionsEnabled && detectedRequest.action !== 'none') {
+                /** @type {DetectedRequest["state"]} */
+                const nextState = { allowed: { reason: 'protectionDisabled' } }
+                if (!nextState) continue // This should never happen, but if it
```
Comment (jonathanKingston) [inline shared/js/background/classes/privacy-dashboard-data.js:70]: This is assigned on the line before; how is it needed?

**Comment by jonathanKingston** [comment]
Overall looks great; I've not tested it yet as you mentioned the search is still broken and the tests are failing still.
The CTP test should disable itself, I reran and was fine.

File: shared/js/background/classes/privacy-dashboard-data.js (hunk 10 above)
Comment (shakyShane) [reply shared/js/background/classes/privacy-dashboard-data.js:72]: yep, it was a leftover

File: shared/js/background/classes/tracker.js [hunk 11]
```diff
@@ -43,22 +29,47 @@ export class Tracker {
...
+
+        // return early if this combination exists.
+        if (this.urls[key]) return
+
+        const state = convertState(t.action, t.sameEntity)
+
+        // if we can't convert the state, do nothing.
+        if (!state) return
+
+        // Chose the first tracker radar category that we accept
```
Comment (jonathanKingston) [inline shared/js/background/classes/tracker.js:56]:
```suggestion
        // Choose the first tracker radar category that we accept
```

File: browsers/chrome-mv3/manifest.json [hunk 12]
```diff
@@ -11,8 +11,8 @@
     "manifest_version": 3,
     "minimum_chrome_version": "102.0",
     "action": {
-        "default_icon": "img/icon_48.png",
```
Comment (jonathanKingston) [inline browsers/chrome-mv3/manifest.json:14]: Seems like this icon can go too.

**Review by jonathanKingston — APPROVED** [review]

**Comment by jonathanKingston** [comment]
Very triggering gif of snape and dumbledoor with the caption 'party hard'

**Pull request merged**